### PR TITLE
Prepare for adding context to the Interner

### PR DESCRIPTION
**Pull request by yaahc**
part of #328

### CODE REVIEW

**Comment by yaahc** [comment]
The current puzzle is how best to access an `&I` from within the various functions on `SlgContext<I>`

```
error[E0061]: this function takes 3 parameters but 2 parameters were supplied
   --> chalk-solve/src/solve/slg.rs:164:37
    |
164 |           let (mut infer, subst, _) = InferenceTable::from_canonical(goal.universes, &goal.canonical);
    |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 parameters
    |
   ::: chalk-solve/src/infer.rs:49:5
    |
49  | /     pub(crate) fn from_canonical<T>(
50  | |         interner: &I,
51  | |         num_universes: usize,
52  | |         canonical: &Canonical<T>,
...   |
67  | |         (table, subst, value)
68  | |     }
    | |_____- defined here

```

File: chalk-integration/src/db.rs [hunk 1]
```diff
@@ -54,7 +54,7 @@ impl ChalkDatabase {
         tls::set_current_program(&program, || op(&program))
     }
 
-    pub fn parse_and_lower_goal(&self, text: &str) -> Result<Goal<ChalkIr>, ChalkError> {
+    pub fn parse_and_lower_goal(&self, interner: &ChalkIr, text: &str) -> Result<Goal<ChalkIr>, ChalkError> {
```
Comment (jackh726) [inline chalk-integration/src/db.rs]: Is this used?

File: chalk-engine/src/context.rs [hunk 2]
```diff
@@ -119,6 +119,8 @@ pub trait Context: Clone + Debug {
     /// the equality relation.
     type Variance;
 
+    type Interner;
```
Comment (jackh726) [inline chalk-engine/src/context.rs]: If you could add a short doc comment, that would be good

File: chalk-integration/src/query.rs [hunk 3]
```diff
@@ -28,7 +28,7 @@ pub trait LoweringDatabase: RustIrDatabase<ChalkIr> {
     #[salsa::input]
     fn solver_choice(&self) -> SolverChoice;
 
-    fn program_ir(&self) -> Result<Arc<Program>, ChalkError>;
+    fn program_ir<'a>(&self, interner: ChalkIr) -> Result<Arc<Program>, ChalkError>;
```
Comment (jackh726) [inline chalk-integration/src/query.rs]: I think this lifetime is uneeded.

File: chalk-ir/src/lib.rs [hunk 4]
```diff
@@ -811,6 +811,12 @@ pub struct AliasEq<I: Interner> {
     pub ty: Ty<I>,
 }
 
+impl<I: Interner> AliasEq<I> {
+    pub fn interner(&self) -> &I {
```
Comment (jackh726) [inline chalk-ir/src/lib.rs]: Yeah, I don't think this function makes sense here.

Comment (yaahc) [reply chalk-ir/src/lib.rs]: It has to either come from the AliasEq or it has to be part of the trait and no other implementor of this trait uses the interner yet.

Comment (jackh726) [reply chalk-ir/src/lib.rs]: @detrumi is working the `impl Trait`/`Alias`, so this portion of the API might change. But regardless, even if the other types don't use it and we still need it, I don't think having `AliasEq` store the interner is correct.

**Pull request merged**